### ISSUE TITLE
Local attributes should override global attributes

### ISSUE THREAD

**Issue by elyisgreat** [alda-lang/alda#199]
I noticed that while the other issue with attributes is fixed, a new one has arisen. I do not know if the recent fix had anything to do with it, but there's an issue with global attributes and voices.

Consider the following code:
```
midi-synth-voice:
(vol! 200/3)
V2: (vol 20) o4 r4 a8 f+ a4 a8 f+ > (quant 50) c+6 c+ d (quant 90) c+4. < a8 b4 a8 d4 d8 d d b4 a d2
V2: (vol 20) r4 a8 f+ a a a > e4 e8 d4 c+4. < a8 b a4 d d b a d e4.
```
You will notice that the local attribute doesn't apply the first time it is used. However the second time it is applied.

When you remove the global attribute it runs as normal: 

```
midi-synth-voice:
V2: (vol 20) o4 r4 a8 f+ a4 a8 f+ > (quant 50) c+6 c+ d (quant 90) c+4. < a8 b4 a8 d4 d8 d d b4 a d2
V2: (vol 20) r4 a8 f+ a a a > e4 e8 d4 c+4. < a8 b a4 d d b a d e4.
```

**Comment by daveyarwood** [comment]
This is expected behavior -- the way global attributes are implemented, they will take precedence over local ones.

I could see it making sense that local attributes should take precedence in cases where you might want all the instruments to have the same tempo, or key signature, or something, *except for one instrument* for which you may want to override that attribute.

This might be a little tricky to implement due to the way global attributes work under the hood -- when you set a global attribute, Alda keeps track of the specific point in the score where it takes place, and then in every instrument part, before each event Alda checks to see if there is a global attribute that needs to be applied at that point in time. We could solve this problem by keeping track of which local attributes were applied at that point in time, and only apply the global attribute if the same attribute wasn't also applied locally at the same time.

**Comment by elyisgreat** [comment]
True, but why would it apply it the second line? 

I was operating under the assumption that local attributes overrode global ones. It seems more logical that way; the global acts as a "default" attribute, with the local ones overriding when necessary.

**Comment by daveyarwood** [comment]
What's happening here is this:

- `(vol! 200/3)` gets evaluated, it adds an entry to the global attributes lookup table that says "at offset 0 (the beginning of the score), set volume to 200/3."
- Then the voice group starts, and the volume gets set to 20.
- Then Alda checks to see if there are any global attributes to be applied before it executes the next event. It sees that at this point in time, there is a global attribute for setting the volume to 200/3, so it sets the volume *back to 200/3*.

The way this is implemented is fairly naïve -- I think we should make it smarter, or at least add a workaround to allow for overriding global attributes with local ones.

> I was operating under the assumption that local attributes overrode global ones. It seems more logical that way; the global acts as a "default" attribute, with the local ones overriding when necessary.

Well put. That's what I was trying to say, you just said it better ;)

**Comment by elyisgreat** [comment]
I think the best way to implement global attributes is as the default values for any new instruments coming in.

 We can leave open for discussion whether or not global attribute calls midway through a score should act as a global attribute set, or if instead they should merely change the default for any new instruments/voices coming in.

**Comment by daveyarwood** [comment]
I think it's better for global attributes to be associated with a moment in time in the score. That way you can concisely change the key signature, time signature, volume, etc. of all instruments midway through the score without having to change it for every instrument.

Global attributes already do the correct thing there, the only issue is that they aren't overridden by local attributes that happen at the same time, whereas they should be.

**Comment by elyisgreat** [comment]
perhaps we could have a logic like this:

-local attributes override global attributes
-global attributes act as the default
-a global attribute call changes the attribute for all instruments not being overridden by a local attribute
-a call such as `(attribute :default)` resets an instrument's local attribute to the global attribute

**Comment by daveyarwood** [comment]
My intention for global attributes isn't for them to be "default values," but rather that they change the attribute for all instruments in the score, instead of just the current instrument.

I think "default value" semantics will be better handled by variables ( #131 ) in the near future. We'll be able to do things like:

```
defaultVolume = (volume 75)
quieter = (volume 50)
louder = (volume 100)

piano:
  defaultVolume c8 d e f g2
  quieter c8 d e f g2
  louder c8 d e f g2
  defaultVolume > c2
```

**Comment by elyisgreat** [comment]
Ah. I realize now that instruments coming in at a certain time makes no sense. all instruments come in at the beginning, thus there's no need for "default" values for when instruments come in. I'm not sure then why they exist at all; rather `(attribute! value)` simply changes the local attributes of all instruments at the time of its call.

**Comment by daveyarwood** [comment]
> `(attribute! value)` simply changes the local attributes of all instruments at the time of its call.

Yep, that's it exactly.

**Comment by daveyarwood** [comment]
This issue is fixed in 1.0.0-rc15.

Now, when local attributes occur at the same point in the score where there is a global attribute, the local attribute will take precedence.

I've added an [example score](https://github.com/alda-lang/alda/blob/master/examples/overriding-a-global-attribute.alda) where there is a global tempo of 60 bpm at the beginning of the score, which the banjo overrides with a local tempo of 180 bpm.
